### PR TITLE
Constrained standard errors

### LINKED ISSUE #57: Robust standard error calculation for constraint estimations and better standard error interface
## Current Situation

- Standard error functions are relatively low level. The user explicitly calculates numerical derivatives and then passes them into the covariance/standard error functions
- It is possible (and happens frequently) that calculation of numerical derivatives fails because the stepsize is so large that some constraints are violated

## Desired Situation

- Standard error calculation can be done in one step
- Numerical derivatives should respect constraints

## Proposed Implementation

Write high level standard error functions that have a similar interface to maximize and minimize. Numerical derivatives are calculated on on a reparametrized version of the objective functions, such that all constraints hold. Then first calculate the covariance matrix for internal_params and use a parametric bootstrap (i.e. monte carlo integration) to calculate the covariance matrix of the external parameters

## Discussion

- Is anybody aware of literature on the topic? Is there any other approach that is similarly easy to apply?
- How would we handle simple box constraints?

### PR DESCRIPTION
**Pull request by janosg**
Currently, we don't support calculation of likelihood standard errors with constraints.  Mainly, because there is no clear and theoretically clean way to do this. This PR adds a very rudimentary version of a likelihood inference function that works with estimagic constraints. The constraints are handled by calculating a covariance on the internal parameter vector and then using monte carlo integration to calculate the covariance matrix of the free but external parameters.

It is clearly experimental at this stage, but having a basic version out means that we can improve on it instead of writing different scripts in each project. 

closes #57

### CODE REVIEW

File: estimagic/inference/shared.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+from estimagic.optimization.process_constraints import process_constraints
+from estimagic.optimization.reparametrize import reparametrize_from_internal
+from estimagic.optimization.reparametrize import reparametrize_to_internal
+
+
+def transform_covariance(
+    params, internal_cov, constraints, n_samples, bounds_handling,
+):
+    """Transform in internal covariance matrix to an external one, given constraints.
```
Comment (roecla) [inline estimagic/inference/shared.py:13]:
```suggestion
    """Transform the internal covariance matrix to an external one, given constraints.
```

File: estimagic/inference/shared.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            transformed = reparametrize_from_internal(
+                internal=params_vec,
+                fixed_values=fixed_values,
+                pre_replacements=pre_replacements,
+                processed_constraints=processed_constraints,
+                post_replacements=post_replacements,
+            )
+            transformed_free.append(transformed[is_free])
+
+            free_cov = np.cov(np.array(transformed_free), rowvar=False,)
```
Comment (roecla) [inline estimagic/inference/shared.py:78]: I think you can indent `free_cov = np.cov(np.array(transformed_free), rowvar=False,)` one level out and just calculate the covariance once after the loop.

**Review by roecla — APPROVED** [review]
Just some very minor comments. Very nice! I think it makes a lot of sense to set up the structure and add a basic version now.

File: estimagic/inference/shared.py (hunk 2 above)
Comment (janosg) [reply estimagic/inference/shared.py]: Done

**Pull request merged**